### ISSUE TITLE
Cannot reflect status to work when retain failed

### ISSUE THREAD

**Issue by Poor12** [karmada-io/karmada#3572]
**What happened**:
Situation 1:  When I deploy a Pod to a Kubernetes cluster which may automatically inject some fields into the Pod (e.g a certain volume is auto mounted like `k8s-api-access`, or add some default resource requests and limit), now Karmada will do the `retain` operation because of the difference of the control plane and member clusters but this may be blocked by admission controller. For Pod, we can only update some limited fields, like image and so on. At this time, the retain failure will cause the control plane to fail to collect the status of the member clusters.

Situation 2:  When I configure the error retain rules, the situation of the above situation will happen, too.

Reference codebase:

    if needUpdate {
            if err := c.ObjectWatcher.Update(clusterName, desiredObj, observedObj); err != nil {
                    klog.Errorf("Update %s failed: %v", fedKey.String(), err)
                    return err
            }
            // We can't return even after a success updates, because that might lose the chance to collect status.
            // Not all updates are real, they might be no change, in that case there will be no more event for this update,
            // this usually happens with those resources not enables 'metadata.generation', like 'Service'.
            // When a Service's status changes, it's 'metadata.resourceVersion' will be increased, but 'metadata.generation'
            // not increased(defaults to 0), the ObjectWatcher can't easily tell what happened to the object, so ObjectWatcher
            // also needs to update again. The update operation will be a non-operation if the event triggered by Service's
            // status changes.
    }

    klog.Infof("reflecting %s(%s/%s) status to Work(%s/%s)", observedObj.GetKind(), observedObj.GetNamespace(), observedObj.GetName(), workNamespace, workName)
    return c.reflectStatus(workObject, observedObj)

**What you expected to happen**:
Even if the `retain` operation failed, Karmada can still collect the status of member clusters and keep them in sync.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Karmada version:
- kubectl-karmada or karmadactl version (the result of `kubectl-karmada version` or `karmadactl version`):
- Others:

**Comment by RainbowMango** [comment]
I'm trying to understand the use case.

> e.g a certain volume is auto mounted like k8s-api-access, or add some default resource requests and limit

Can you give some more details, like some links to relevant documents? I'm thinking if it is possible to apply the injection on Karmada control plane.

**Comment by MingZhang-YBPS** [comment]
I met the same issue as well. Generally it means user is not aware of failure when the update operation is rejected by admission webhook on member cluster.

1. A object is created in Karmada and scheduled to a member cluster successfully
2. User update some fields of this object on Karmada, and hope these changes to be propagated to member cluster
3. But the update operation launched by execution controller may fail on member cluster because of the admission webhook error, then execution controller simply retry again and again
4. User won't find any exception by inspecting the object on Karmada

**Comment by XiShanYongYe-Chang** [comment]
This problem essentially feels like the same issue： #3587 

The root cause is that there was an error in synchronizing resources with the member cluster during state collection, which prevented the continuation of the state collection operation and resulted in the inability to update the state properly.

**Comment by XiShanYongYe-Chang** [comment]
We can try to consider continuing to collect the status after synchronization failure.

Hi @MingZhang-YBPS, would like to have a try?

**Comment by MingZhang-YBPS** [comment]
> We can try to consider continuing to collect the status after synchronization failure.
> 
> Hi @MingZhang-YBPS, would like to have a try?

Thanks, it is a good chance to contribute. And let me try to summary the possible solution as below.

The problem: User may update his object fields on Karmada control plane, and he simply wants these change to be propagated to member clusters. But object fields may not be allowed to be updated into member clusters due to admission error. Besides that, #3587 is another simliar case. 
Karmada should provide **general way to present out-of-sync error** to User. Can we enhance it like below?

1. Each time, when the object fields is changed on Karmada control plane, its generation/resourceversion is supposed to be updated as well. 
2. After Karmada execution controller is trying to applied to member clusters, it will set two new annotation _work.karmada.io/last-attempted-revision_ and _work.karmada.io/last-applied-revision_ as well. The second one is the revision which Karmada execution controller applied successfully.
3. This approach is similar with how fluxcd is designed https://fluxcd.io/flux/components/kustomize/kustomization/#last-applied-revision and Karmada handles it https://github.com/karmada-io/karmada/blob/master/pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/kustomize.toolkit.fluxcd.io/v1/Kustomization/customizations.yaml#L33-L38
4. If user wants to know the exact result, he can get it from CRB/RB condition FullyAppliedSuccess or FullyAppliedFailed which is already reflected.
![image](https://github.com/karmada-io/karmada/assets/74715700/d9ee10ef-628f-4da8-aa5c-946fc7cd56fb)
5. I don't prefer to add these new field in object status directly, since there maybe some objects which don't have status at all.

**Comment by XiShanYongYe-Chang** [comment]
Hi @MingZhang-YBPS Thanks for your summary and resolve.

I'm sorry, I didn't explain the problem clearly. The synchronization of the modified resource template on the control plane is handled by the execution-controller. If there is a synchronization failure, it will be indicated in the condition of the works status. So far, this process seems to be working fine.

The current issue lies in the workstatus controller. It checks for any modifications to the resources on the member cluster during the process of collecting status. If changes occur, it needs to continue with synchronization. However, if an error occurs during the synchronization process, it will interrupt the collection of work status.

I don't know if I have expressed this question clearly. If you have any doubts, feel free to continue raising them.

**Comment by MingZhang-YBPS** [comment]
> Hi @MingZhang-YBPS Thanks for your summary and resolve.
> 
> I'm sorry, I didn't explain the problem clearly. The synchronization of the modified resource template on the control plane is handled by the execution-controller. If there is a synchronization failure, it will be indicated in the condition of the works status. So far, this process seems to be working fine.
> 
> The current issue lies in the workstatus controller. It checks for any modifications to the resources on the member cluster during the process of collecting status. If changes occur, it needs to continue with synchronization. However, if an error occurs during the synchronization process, it will interrupt the collection of work status.
> 
> I don't know if I have expressed this question clearly. If you have any doubts, feel free to continue raising them.

Umm, they are two different issues
1. If ObjectWatcher.Update fails in workstatus controller, the object status between Karmada control plane and member cluster goes out-of-sync. That's a bug to be fixed, I can do it.
2. How user aware his change not applied to member cluster? It looks like he has to check the RB/CRB, not resource template directly

**Comment by XiShanYongYe-Chang** [comment]
OK, thanks

**Comment by lxtywypc** [comment]
In my opinion, we had better not continue reflecting status when `ObjectWatcher.Update` failed but set the `Applied` `condition` in  `work status` to `False`, so that the `FullyApplied` `condition` in `binding status` could be aggregated as `False`, and the `aggregatedStatus` would also contain an item with `applied = False` which info could be processed by `resource interpreter` to reflect the situation into object status on `fed control palne`.

**Comment by XiShanYongYe-Chang** [comment]
> In my opinion, we had better not continue reflecting status when `ObjectWatcher.Update` failed but set the `Applied` `condition` in `work status` to `False`, so that the `FullyApplied` `condition` in `binding status` could be aggregated as `False`, and the `aggregatedStatus` would also contain an item with `applied = False` which info could be processed by `resource interpreter` to reflect the situation into object status on `fed control palne`.

Based on your modifications #3808 , does this problem no longer exist?

**Comment by lxtywypc** [comment]
> Based on your modifications https://github.com/karmada-io/karmada/pull/3808 , does this problem no longer exist?  

Here is a test flow:  

1. Firstly, we deployed a `deployment` named `wy-test` in `test` namespace and create a `propagationpolicy` for it to distribute it into `cluster1` and `cluster2` as following:  

```yaml
# deployment
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    work.karmada.io/conflict-resolution: overwrite
  labels:
    app: wy-test
  name: wy-test
  namespace: test
spec:
  replicas: 1
  selector:
    matchLabels:
      app: wy-test-pod
  template:
    metadata:
      labels:
        app: wy-test-pod
      name: wy-test-pod
      namespace: test
    # ...
```

```yaml
# propagationpolicy
apiVersion: policy.karmada.io/v1alpha1
kind: PropagationPolicy
metadata:
  name: test
  namespace: test
spec:
  placement:
    clusterAffinity:
      clusterNames:
      - cluster1
      - cluster2
  resourceSelectors:
  - apiVersion: apps/v1
    kind: Deployment
    namespace: test
```

After a while, we could get the `work` for `cluster1` like:  

```
[fed-control-plane]$ kubectl get work -n karmada-es-cluster1  wy-test-cb9f6b8d9 -o yaml
apiVersion: work.karmada.io/v1alpha1
kind: Work
metadata:
  annotations:
    resourcebinding.karmada.io/name: wy-test-deployment
    resourcebinding.karmada.io/namespace: test
  labels:
    resourcebinding.karmada.io/key: 54596fb559
  name: wy-test-cb9f6b8d9
  namespace: karmada-es-cluster1
spec:
  workload:
    manifests:
    - apiVersion: apps/v1
      kind: Deployment
      metadata:
        annotations:
          resourcebinding.karmada.io/name: wy-test-deployment
          resourcebinding.karmada.io/namespace: test
          work.karmada.io/conflict-resolution: overwrite
        labels:
          app: wy-test
          propagationpolicy.karmada.io/name: test
          propagationpolicy.karmada.io/namespace: test
          resourcebinding.karmada.io/key: 54596fb559
          work.karmada.io/name: wy-test-cb9f6b8d9
          work.karmada.io/namespace: karmada-es-cluster1
        name: wy-test
        namespace: test
      spec:
      # ...
status:
  conditions:
  - lastTransitionTime: "2023-07-19T07:40:51Z"
    message: Manifest has been successfully applied
    reason: AppliedSuccessful
    status: "True"
    type: Applied
  manifestStatuses:
  - health: Healthy
    identifier:
      group: apps
      kind: Deployment
      name: wy-test
      namespace: test
      ordinal: 0
      resource: ""
      version: v1
    status:
      availableReplicas: 1
      readyReplicas: 1
      replicas: 1
      updatedReplicas: 1
```

2. Then we applied a fake `resourceinterpreterwebhookconfiguration` as following, that would make `retain` failed because we didn't deploy the service:  

```yaml
# resourceinterpreterwebhookconfiguration
apiVersion: config.karmada.io/v1alpha1
kind: ResourceInterpreterWebhookConfiguration
metadata:
  name: deploy
webhooks:
  - name: deployment.k8s.io
    rules:
      - operations:
        - "Retain"
        apiGroups:
        - "apps"
        apiVersions:
        - "v1"
        kinds:
        - "Deployment"
    clientConfig:
      service:
        name: deploy
        namespace: karmada-system
        path: /interpret-deploy
        port: 443
    interpreterContextVersions:
    - "v1alpha1"
    timeoutSeconds: 3
```

Then we tried to scale the `deployment` in `cluster1`:  

```
[cluster1]$ kubectl get deploy -n test
NAME      READY   UP-TO-DATE   AVAILABLE   AGE
wy-test   1/1     1            1           5d1h
[cluster1]$ kubectl scale deploy -n test wy-test --replicas=2
deployment.apps/wy-test scaled
[cluster1]$ kubectl get deploy -n test
NAME      READY   UP-TO-DATE   AVAILABLE   AGE
wy-test   2/2     2            2           5d1h
```

And then we would get the `work` like:  

```
[fed-control-plane]$ kubectl get work -n karmada-es-cluster1 wy-test-cb9f6b8d9 -o yaml
apiVersion: work.karmada.io/v1alpha1
kind: Work
metadata:
  annotations:
    resourcebinding.karmada.io/name: wy-test-deployment
    resourcebinding.karmada.io/namespace: test
  labels:
    resourcebinding.karmada.io/key: 54596fb559
  name: wy-test-cb9f6b8d9
  namespace: karmada-es-cluster1
spec:
  workload:
    manifests:
    - apiVersion: apps/v1
      kind: Deployment
      metadata:
        annotations:
          resourcebinding.karmada.io/name: wy-test-deployment
          resourcebinding.karmada.io/namespace: test
          work.karmada.io/conflict-resolution: overwrite
        labels:
          app: wy-test
          propagationpolicy.karmada.io/name: test
          propagationpolicy.karmada.io/namespace: test
          resourcebinding.karmada.io/key: 54596fb559
          work.karmada.io/name: wy-test-cb9f6b8d9
          work.karmada.io/namespace: karmada-es-cluster1
        name: wy-test
        namespace: test
      spec:
      # ...
status:
  conditions:
  - lastTransitionTime: "2023-07-19T07:40:45Z"
    message: Manifest has been successfully applied
    reason: AppliedSuccessful
    status: "True"
    type: Applied
  manifestStatuses:
  - health: Healthy
    identifier:
      group: apps
      kind: Deployment
      name: wy-test
      namespace: test
      ordinal: 0
      resource: ""
      version: v1
    status:
      availableReplicas: 1
      readyReplicas: 1
      replicas: 1
      updatedReplicas: 1
```

Nothing changed, status not synced, and we got logs from `karmada-controller-manager`:  

```
I0719 10:34:37.475914       1 workstatus_controller.go:175] Begin to sync work status for object(cluster=cluster1, apps/v1, kind=Deployment, test/wy-test)
I0719 10:34:37.476513       1 interpreter.go:135] Begin to retain object: apps/v1, Kind=Deployment test/wy-test.
E0719 10:34:37.477588       1 objectwatcher.go:163] Failed to retain fields for resource(kind=Deployment, test/wy-test) in cluster cluster1: Internal error occurred: failed calling webhook "deploy/deployment.k8s.io": failed to call webhook: Post "https://deploy.karmada-system.svc:443/interpret-deploy?timeout=3s": service "deploy" not found
E0719 10:34:37.477613       1 workstatus_controller.go:229] Update cluster=cluster1, apps/v1, kind=Deployment, test/wy-test failed: Internal error occurred: failed calling webhook "deploy/deployment.k8s.io": failed to call webhook: Post "https://deploy.karmada-system.svc:443/interpret-deploy?timeout=3s": service "deploy" not found
```

3. Then we changed the image of `karmada-controller-manager` to apply the changes in #3808, removed the `resourceinterpreterwebhookconfiguration`, scaled the `deployment` in fed control plane and waited for status synced:  

```
[fed-control-plane]$ kubectl scale deploy -n test wy-test --replicas=3
deployment.apps/wy-test scaled
[fed-control-plane]$ kubectl get work -n karmada-es-cluster1 wy-test-cb9f6b8d9 -o yaml
apiVersion: work.karmada.io/v1alpha1
kind: Work
metadata:
  annotations:
    resourcebinding.karmada.io/name: wy-test-deployment
    resourcebinding.karmada.io/namespace: test
  labels:
    resourcebinding.karmada.io/key: 54596fb559
  name: wy-test-cb9f6b8d9
  namespace: karmada-es-cluster1
spec:
  workload:
    manifests:
    - apiVersion: apps/v1
      kind: Deployment
      metadata:
        annotations:
          resourcebinding.karmada.io/name: wy-test-deployment
          resourcebinding.karmada.io/namespace: test
          work.karmada.io/conflict-resolution: overwrite
        labels:
          app: wy-test
          propagationpolicy.karmada.io/name: test
          propagationpolicy.karmada.io/namespace: test
          resourcebinding.karmada.io/key: 54596fb559
          work.karmada.io/name: wy-test-cb9f6b8d9
          work.karmada.io/namespace: karmada-es-cluster1
        name: wy-test
        namespace: test
      spec:
      # ...
status:
  conditions:
  - lastTransitionTime: "2023-07-19T10:48:51Z"
    message: Manifest has been successfully applied
    reason: AppliedSuccessful
    status: "True"
    type: Applied
  manifestStatuses:
  - health: Healthy
    identifier:
      group: apps
      kind: Deployment
      name: wy-test
      namespace: test
      ordinal: 0
      resource: ""
      version: v1
    status:
      availableReplicas: 3
      readyReplicas: 3
      replicas: 3
      updatedReplicas: 3
```

4. Then we applied the fake `resourceinterpreterwebhookconfiguration` and scaled the `deployment` in `cluster1` again, this time we go the `work` like this:  

```
[cluster1]$ kubectl get deploy -n test
NAME      READY   UP-TO-DATE   AVAILABLE   AGE
wy-test   3/3     3            3           5d1h
[cluster1]$ kubectl scale deploy -n test wy-test --replicas=2
deployment.apps/wy-test scaled
[cluster1]$ kubectl get deploy -n test
NAME      READY   UP-TO-DATE   AVAILABLE   AGE
wy-test   2/2     2            2           5d1h
```

```
[fed-control-plane]$ kubectl get work -n karmada-es-cluster1   wy-test-cb9f6b8d9 -o yaml
apiVersion: work.karmada.io/v1alpha1
kind: Work
metadata:
  annotations:
    resourcebinding.karmada.io/name: wy-test-deployment
    resourcebinding.karmada.io/namespace: test
  labels:
    resourcebinding.karmada.io/key: 54596fb559
  name: wy-test-cb9f6b8d9
  namespace: karmada-es-cluster1
spec:
  workload:
    manifests:
    - apiVersion: apps/v1
      kind: Deployment
      metadata:
        annotations:
          resourcebinding.karmada.io/name: wy-test-deployment
          resourcebinding.karmada.io/namespace: test
          work.karmada.io/conflict-resolution: overwrite
        labels:
          app: wy-test
          propagationpolicy.karmada.io/name: test
          propagationpolicy.karmada.io/namespace: test
          resourcebinding.karmada.io/key: 54596fb559
          work.karmada.io/name: wy-test-cb9f6b8d9
          work.karmada.io/namespace: karmada-es-cluster1
        name: wy-test
        namespace: test
      spec:
      # ...
status:
  conditions:
  - lastTransitionTime: "2023-07-19T10:51:10Z"
    message: 'Failed to apply all manifests (0/1): Internal error occurred: failed
      calling webhook "deploy/deployment.k8s.io": failed to call webhook: Post "https://deploy.karmada-system.svc:443/interpret-deploy?timeout=3s":
      service "deploy" not found'
    reason: AppliedFailed
    status: "False"
    type: Applied
  manifestStatuses:
  - health: Unhealthy
    identifier:
      group: apps
      kind: Deployment
      name: wy-test
      namespace: test
      ordinal: 0
      resource: ""
      version: v1
    status:
      availableReplicas: 3
      readyReplicas: 3
      replicas: 3
      updatedReplicas: 3
```

I have cut some info unimportant to make it simple.  

It seemed to go on as what I want.

**Comment by XiShanYongYe-Chang** [comment]
Thank you for your brilliant explanation, so we can conclude that #3808 has already resolved the current issue as well as #3587 's problem.
